src/components/Analyse/components/SegmentPositionContext.tsx

```diff
@@ -0,0 +1,72 @@
+import React from 'react'
+
+export interface Scrollable {
+	scrollIntoView (): void
+}
+
+export interface Context {
+	active: number|null
+	register (scrollable: Scrollable, id: number, state: boolean): void
+	unregister (id: number): void
+	scrollToId (id?: number): void
+}
+
+const context = React.createContext<Context>(undefined as never)
+export const { Consumer } = context
+
+interface ProviderState extends Context {
+	registry: ReadonlyMap<number, boolean>
+}
+
+export class SegmentPositionProvider extends React.PureComponent<{}, ProviderState> {
+	readonly state: Readonly<ProviderState> = {
+		active: null,
+		register: this.register.bind(this),
+		unregister: this.unregister.bind(this),
+		scrollToId: this.scrollToId.bind(this),
+		registry: new Map()
+	}
+	private readonly refMap = new Map<number, { scrollIntoView (): void }>()
+
+	componentDidUpdate (_prevProps: Readonly<{}>, prevState: Readonly<ProviderState>) {
+		const { registry } = this.state
+		if (registry !== prevState.registry) {
+			const toCheck = Array.from(registry.keys()).sort()
+			const activeId = toCheck.find(id => registry.get(id)!)
+			const active = activeId !== undefined ? activeId : (
+				toCheck.length > 0 ? toCheck[toCheck.length - 1] : null
+			)
+
+			if (active !== this.state.active) {
+				this.setState({ active })
+			}
+		}
+	}
+
+	render () {
+		return <context.Provider value={this.state}>{this.props.children}</context.Provider>
```

Comment:
Using `this.state` is kind of a dirty approach but avoids creating a new object on every render / having to have a separate memoized context object.